### ISSUE TITLE
Propose an RFC to reduce the overhead caused by `if-else` statement in the forward pass of all operators.

### ISSUE THREAD

**Issue by vllmellm** [EmbeddedLLM/vllm#38]
vLLM's V0 architecture does not use CUDA graphs for the prefill stage. Prefill attention is not compatible with cudagraphs in V0. 

Thus, we would like to propose that the operators should be abstracted into a module abstraction like PyTorch module to reduce the if-else condition in the actual forward pass of the function especially those backend that still depends on V0 (V1 needs to be investigated). There are cases that we found that if-else condition is increasing the latency, e.g. like in the MLAAttention Case:

https://github.com/vllm-project/vllm/pull/15893/commits/02a4fb321a4d67f67310cd91cb6678018670774c 

## Slower Implementation
```
class AiterMLAImpl(MLACommonImpl[AiterMLAMetadata]):
    def __init__(
            self,
            num_heads: int,
            head_size: int,
            scale: float,
            num_kv_heads: int,
            alibi_slopes: Optional[list[float]],
            sliding_window: Optional[int],
            kv_cache_dtype: str,
            blocksparse_params: Optional[dict[str, Any]],
            logits_soft_cap: Optional[float],
            attn_type: str,
            # MLA Specific Arguments
            **mla_args) -> None:
        super().__init__(num_heads, head_size, scale, num_kv_heads,
                         alibi_slopes, sliding_window, kv_cache_dtype,
                         blocksparse_params, logits_soft_cap, attn_type,
                         **mla_args)
        unsupported_features = [
            alibi_slopes, sliding_window, blocksparse_params, logits_soft_cap
        ]
        if any(unsupported_features):
            raise NotImplementedError(
                "Aiter MLA does not support one of the following: "
                "alibi_slopes, sliding_window, blocksparse_params, "
                "logits_soft_cap")

        flash_attn_func = None
        with suppress(ImportError):
            if is_aiter_fa_enabled():
                from aiter import flash_attn_varlen_func
                flash_attn_func = flash_attn_varlen_func
            else:
                from flash_attn import flash_attn_varlen_func
                flash_attn_func = flash_attn_varlen_func

        self.flash_attn_varlen_func = flash_attn_func

    def _get_fwd_prefill_attn_output(self, q: torch.Tensor, k: torch.Tensor,
                                     v: torch.Tensor,
                                     prefill_metadata: MLACommonMetadata,
                                     kv_c_and_k_pe_cache: torch.Tensor,
                                     attn_metadata: MLACommonMetadata,
                                     has_context: bool) -> torch.Tensor:
        if is_aiter_fa_enabled():
            is_v_padded = False
            output = self.flash_attn_varlen_func(
                q=q,
                k=k,
                v=v,
                cu_seqlens_q=prefill_metadata.query_start_loc,
                cu_seqlens_k=prefill_metadata.query_start_loc,
                max_seqlen_q=prefill_metadata.max_prefill_seq_len,
                max_seqlen_k=prefill_metadata.max_prefill_seq_len,
                softmax_scale=self.scale,
                causal=True,
                return_lse=has_context,
            )

        else:
            # For MLA the v head dim is smaller than qk head dim so we pad out
            # v with 0s to match the qk head dim
            v_padded = torch.nn.functional.pad(v,
                                               [0, q.shape[-1] - v.shape[-1]],
                                               value=0)
            is_v_padded = True
            output = self.flash_attn_varlen_func(
                q=q,
                k=k,
                v=v_padded,
                cu_seqlens_q=prefill_metadata.query_start_loc,
                cu_seqlens_k=prefill_metadata.query_start_loc,
                max_seqlen_q=prefill_metadata.max_prefill_seq_len,
                max_seqlen_k=prefill_metadata.max_prefill_seq_len,
                softmax_scale=self.scale,
                causal=True,
                return_attn_probs=has_context,
            )

        if has_context:
            # ROCm flash_attn_varlen_func will return 3 objects instead of 2
            suffix_output, suffix_lse, *rest = output
            context_output, context_lse = self._compute_prefill_context( \
                q, kv_c_and_k_pe_cache, attn_metadata)

            output = torch.empty_like(suffix_output)
            merge_attn_states(
                output=output,
                prefix_output=context_output,
                prefix_lse=context_lse,
                suffix_output=suffix_output,
                suffix_lse=suffix_lse,
            )

        if is_v_padded:
            # slice by `:v.shape[-1]` in order to remove v headdim padding
            return output\
                .view(-1, self.num_heads, q.shape[-1])[..., :v.shape[-1]]\
                    .reshape(-1, self.num_heads * v.shape[-1])
        else:
            return output.reshape(-1, self.num_heads * v.shape[-1])

    def _get_prefill_ctx_attn_output(
            self, index: int, q: torch.Tensor, k: torch.Tensor,
            v: torch.Tensor,
            metadata: MLACommonMetadata) -> Tuple[torch.Tensor, ...]:
        assert metadata.context_chunk_cu_seq_lens is not None
        assert metadata.context_chunk_max_seq_lens is not None

        if is_aiter_fa_enabled():
            return self.flash_attn_varlen_func(
                q=q,
                k=k,
                v=v,
                cu_seqlens_q=metadata.query_start_loc,
                cu_seqlens_k=metadata.context_chunk_cu_seq_lens[index],
                max_seqlen_q=metadata.max_query_len,
                max_seqlen_k=metadata.context_chunk_max_seq_lens[index],
                softmax_scale=self.scale,
                causal=False,  # Context is unmasked
                return_lse=True,
            )
        else:
            attn_output, attn_softmax_lse, _ = self.flash_attn_varlen_func(
                q=q,
                k=k,
                v=v,
                cu_seqlens_q=metadata.query_start_loc,
                cu_seqlens_k=metadata.context_chunk_cu_seq_lens[index],
                max_seqlen_q=metadata.max_query_len,
                max_seqlen_k=metadata.context_chunk_max_seq_lens[index],
                softmax_scale=self.scale,
                causal=False,  # Context is unmasked
                return_attn_probs=True,
            )
            return attn_output, attn_softmax_lse

```

## Faster Implementation (with if-else statement removed from the prefill)

```
class AiterMLAImpl(MLACommonImpl[AiterMLAMetadata]):

    def __init__(
            self,
            num_heads: int,
            head_size: int,
            scale: float,
            num_kv_heads: int,
            alibi_slopes: Optional[list[float]],
            sliding_window: Optional[int],
            kv_cache_dtype: str,
            blocksparse_params: Optional[dict[str, Any]],
            logits_soft_cap: Optional[float],
            attn_type: str,
            # MLA Specific Arguments
            **mla_args) -> None:
        super().__init__(num_heads, head_size, scale, num_kv_heads,
                         alibi_slopes, sliding_window, kv_cache_dtype,
                         blocksparse_params, logits_soft_cap, attn_type,
                         **mla_args)

        unsupported_features = [
            alibi_slopes, sliding_window, blocksparse_params, logits_soft_cap
        ]
        if any(unsupported_features):
            raise NotImplementedError(
                "Aiter MLA does not support one of the following: "
                "alibi_slopes, sliding_window, blocksparse_params, "
                "logits_soft_cap")

        from aiter import flash_attn_varlen_func
        self.flash_attn_varlen_func = flash_attn_varlen_func

    def _get_fwd_prefill_attn_output(self, q: torch.Tensor, k: torch.Tensor,
                                     v: torch.Tensor,
                                     prefill_metadata: MLACommonMetadata,
                                     kv_c_and_k_pe_cache: torch.Tensor,
                                     attn_metadata: MLACommonMetadata,
                                     has_context: bool) -> torch.Tensor:
        output = self.flash_attn_varlen_func(
            q=q,
            k=k,
            v=v,
            cu_seqlens_q=prefill_metadata.query_start_loc,
            cu_seqlens_k=prefill_metadata.query_start_loc,
            max_seqlen_q=prefill_metadata.max_prefill_seq_len,
            max_seqlen_k=prefill_metadata.max_prefill_seq_len,
            softmax_scale=self.scale,
            causal=True,
            return_lse=has_context,
        )

        if has_context:
            suffix_output, suffix_lse = output
            context_output, context_lse = self._compute_prefill_context( \
                q, kv_c_and_k_pe_cache, attn_metadata)

            output = torch.empty_like(suffix_output)
            merge_attn_states(
                output=output,
                prefix_output=context_output,
                prefix_lse=context_lse,
                suffix_output=suffix_output,
                suffix_lse=suffix_lse,
            )

        return output.reshape(-1, self.num_heads * v.shape[-1])

    def _get_prefill_ctx_attn_output(
            self, index: int, q: torch.Tensor, k: torch.Tensor,
            v: torch.Tensor,
            metadata: MLACommonMetadata) -> Tuple[torch.Tensor, ...]:
        assert metadata.context_chunk_cu_seq_lens is not None
        assert metadata.context_chunk_max_seq_lens is not None

        return self.flash_attn_varlen_func(
            q=q,
            k=k,
            v=v,
            cu_seqlens_q=metadata.query_start_loc,
            cu_seqlens_k=metadata.context_chunk_cu_seq_lens[index],
            max_seqlen_q=metadata.max_query_len,
            max_seqlen_k=metadata.context_chunk_max_seq_lens[index],
            softmax_scale=self.scale,
            causal=False,  # Context is unmasked
            return_lse=True,
        )
```

| Metric | With AITER FA | Without AITER FA | Removed if-else statement (With AITER FA) |
|--------|---------------|------------------|---------------------------|
| **General Performance** |
| Successful requests | 1000 | 1000 | 1000 |
| Benchmark duration (s) | 123.52 | 126.49 | 105.34 |
| Total input tokens | 1024000 | 1024000 | 1024000 |
| Total generated tokens | 40430 | 41126 | 41491 |
| Request throughput (req/s) | 8.10 | 7.91 | 9.49 |
| Output token throughput (tok/s) | 327.32 | 325.14 | 393.86 |
| Total Token throughput (tok/s) | 8617.67 | 8420.73 | 10114.36 |
| **Time to First Token** |
| Mean TTFT (ms) | 60044.23 | 53207.80 | 46229.36 |
| Median TTFT (ms) | 58099.73 | 47137.27 | 45586.21 |
| P99 TTFT (ms) | 115183.29 | 118779.75 | 97573.51 |
| **Time per Output Token (excl. 1st token)** |
| Mean TPOT (ms) | 3207.22 | 2434.73 | 2350.56 |
| Median TPOT (ms) | 856.59 | 772.80 | 841.96 |
| P99 TPOT (ms) | 26681.00 | 16547.11 | 18355.04 |
| **Inter-token Latency** |
| Mean ITL (ms) | 547.62 | 580.84 | 464.36 |
| Median ITL (ms) | 111.58 | 107.50 | 107.56 |
| P99 ITL (ms) | 8068.61 | 8923.07 | 6253.14 |

**Comment by vllmellm** [comment]
This is issue caused because of bad use of module imports
